### PR TITLE
Adopt Punycode in HttpUrl on non-JVM platforms

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: okhttp/src/commonMain/kotlin/okhttp3/internal/-HostnamesCommon.kt [hunk 1]
```diff
@@ -292,22 +292,18 @@ internal fun String.toCanonicalHost(): String? {
...
-      null
-    } else {
-      result
-    }
-  } catch (_: IllegalArgumentException) {
-    return null
+  val result = idnToAscii(host) ?: return null
+  if (result.isEmpty()) return null
+
+  return if (result.containsInvalidHostnameAsciiCodes()) {
```
Comment (yschimke) [inline okhttp/src/commonMain/kotlin/okhttp3/internal/-HostnamesCommon.kt:298]: Are these two check needed?

Or should be postconditions via check?

File: okhttp/src/commonTest/kotlin/okhttp3/HttpUrlTest.kt [hunk 2]
```diff
@@ -471,14 +471,12 @@ open class HttpUrlTest {
 
   @Test
   fun hostnameLowercaseCharactersMappedDirectly() {
-    if (!isJvm) return // TODO: this test is broken on non-JVM platforms.
```
Comment (yschimke) [inline okhttp/src/commonTest/kotlin/okhttp3/HttpUrlTest.kt:474]: Nice!

**Review by yschimke — APPROVED** [review]
LGTM

Take it with a grain of salt, few beers in watching arsenal game with my son.

File: okhttp/src/commonMain/kotlin/okhttp3/internal/-HostnamesCommon.kt (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/commonMain/kotlin/okhttp3/internal/-HostnamesCommon.kt:298]: Yeah, they’re needed. Java’s IDN ToAscii returns non-ascii characters if it doesn’t have a mapping!

**Pull request merged**